src/benchmarks/config_reader_win.c

```diff
@@ -0,0 +1,333 @@
+#define KEY_GROUP	TEXT("group")
+
+
+/*
+ * Maximum section size according to MSDN documentation
+ */
+#define SIZEOF_SECTION 32767
+
+#define NULL_LIST_EMPTY(x) (_tcslen(x) == 0)
+#define NULL_LIST_NEXT(x) x += (_tcslen(x) + 1)
+
+#define KV_LIST_EMPTY(x) (_tcslen(x) == 0)
+#define KV_FIRST(x)
+#define KV_LIST_NEXT(x) x += (_tcslen(x) + 1), x += (_tcslen(x) + 1), \
+		x = kv_list_skip_comment(x)
+
+#define KV_LIST_KEY(x) (x)
+#define KV_LIST_VALUE(x) (x + _tcslen(x) + 1)
+#define KV_LIST_INIT(x) kv_list_init(x)
+
+#define LIST LPTSTR
+#define KV_LIST LPTSTR
+
+static KV_LIST
+kv_list_skip_comment(KV_LIST list)
+{
+	while (list[0] == TEXT('#'))
+		list += (_tcslen(list) + 1);
+	return list;
+}
+
+static KV_LIST
+kv_list_init(LPTSTR list)
+{
+	list = kv_list_skip_comment(list);
+
+	for (KV_LIST it = list; !KV_LIST_EMPTY(it); KV_LIST_NEXT(it)) {
+		LPTSTR c = _tcsstr(it, TEXT("="));
+		if (c == NULL)
+			return NULL;
+
+		*c = TEXT('\0');
+	}
+	return list;
+}
+
+
+/*
+ * config_reader -- handle structure
+ */
+struct config_reader
+{
+	LPTSTR lpFileName;
+};
+
+/*
+ * config_reader_alloc -- allocate config reader
+ */
+struct config_reader *
+config_reader_alloc(void)
+{
+	struct config_reader *cr = (struct config_reader *)malloc(sizeof(*cr));
+	if (cr == NULL)
+		return NULL;
+
+	return cr;
+}
+/*
+ * config_reader_read -- read config file
+ */
+int
+config_reader_read(struct config_reader *cr, LPCTSTR fname)
+{
+	DWORD len = 0;
+	LPTSTR buf = TEXT(" ");
+	/* get the length of the full pathname incl. terminating null char */
+	len = GetFullPathName(fname, 0, buf, NULL);
+	if (len == 0) {
+		/* the function failed */
+		return -1;
+	} else {
+		/* allocate a buffer large enough to store the pathname */
+		/* XXX: consider using Malloc/Free */
+		LPTSTR buffer = (LPTSTR)malloc(len * sizeof(LPTSTR));
+		DWORD ret = GetFullPathName(fname, len, buffer, NULL);
+
+		if (_taccess(buffer, 0) != 0) {
+			printf("%s", strerror(errno));
+			return -1;
+		}
+		cr->lpFileName = (LPTSTR)buffer;
+	}
+
+
+	return 0;
+}
+
+/*
+ * config_reader_free -- free config reader
+ */
+void
+config_reader_free(struct config_reader *cr)
+{
+	free(cr);
+}
+
+/*
+ * is_scenario -- (internal) return true if _name_ is scenario name
+ *
+ * This filters out the _global_ and _config_ sections.
+ */
+static int
+is_scenario(LPTSTR name)
+{
+	return _tcscmp(name, SECTION_GLOBAL);
+}
+
+/*
+ * is_argument -- (internal) return true if _name_ is argument name
+ *
+ * This filters out the _benchmark_ key.
+ */
+static int
+is_argument(LPTSTR name)
+{
+	return _tcscmp(name, KEY_BENCHMARK) != 0 &&
+		_tcscmp(name, KEY_GROUP) != 0;
+}
+
+/*
+ * config_reader_get_scenarios -- return scenarios from config file
+ *
+ * This function reads the config file and returns a list of scenarios.
+ * Each scenario contains a list of key/value arguments.
+ * The scenario's arguments are merged with arguments from global section.
+ */
+
+int
+config_reader_get_scenarios(struct config_reader *cr,
```

Comment:
I don't like this. We should unify the config parser, not create two separate parsers that very similar yet somewhat different. This creates a maintenance burden. Use a portable library instead of this...